### PR TITLE
Adding RLE-core module for JPEG encoder

### PR DESCRIPTION
**Pull request by ishan98**
This includes the addition of **RLE module to the JPEG encoder** with Test Bench : 

**RLE module** : 
The module takes an input_data and generate an output compressing all the zeros within two non-zero values in the form of runlength, which is used to compress the image.
**RLEmain.py** : 
Connect the input to both the RLEcore and the Entrophycoder.
**RLEcore.py** : 
Takes the input and calculate the runlength that is number of zeros between two non-zero values.
**Entrophycoder.py** : 
Takes the input of the non-zero amplitude and calculate the number of bits required to store that amplitude.

**Test Bench** : 
new_rlemain_tb.py
new_entrophycoder_tb.py
new_rlecore_tb.py
This three modules are made for the testing of the RLEmain, Entrophycoder and RLEcore respectively.

**Common.py** : 
Added input_data for testing the RLE module and various functions such as setdata() to get the input and output in the required format.

After adding the above changes, the resultant repository contains the **working model of the RLE module along with the test suite**.

### CODE REVIEW

File: litejpeg/core/rle/entropycoder.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
+# Entropycoder Module
+
+# Importing libraries.
+from litex.gen import *
+from litex.soc.interconnect.stream import *
+
+from litejpeg.core.common import *
+
+"""
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: This comment should be the first thing in the file. Can you move it there?

File: litejpeg/core/rle/entropycoder.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    """
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the Entropycoder to calculate the bits for the
+    Amplitude to store.
+
+    Parameters:
+    -----------
+    sink : Get the input from the other modules.
+           The input is the modulus of the input.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: The input is the modulus of the input? I'm not sure what that means?

File: litejpeg/core/rle/entropycoder.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        # The values for storing the temporary values for
+        # calculating the size.
+        get_data = Array(Signal(12) for i in range(12))
+
+        # Store the output.
+        size = Signal(4)
+
+        # For calculating the size.
+
+        # Storing the size in the input_data.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py:54]: Your comments should concentrate on *why* you are doing something. The code explains how you are doing it.

For example, why do you want to iterate over different values from 1 to 12? Any python programmer understands that `for i in range(12)` iterates over the values 0->11.

File: litejpeg/core/rle/entropycoder.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        # Connecting the source.data with the output.
+        self.comb += source.data.eq(size)
+
+
+class Entropycoder(PipelinedActor, Module):
+    """
+    This module will connect the Entropycoder datapath with the input
+    and output either from other modules or from the Test Benches.
+    The input is been taken from the sink and source and is been transferred to
+    the Entropycoder datapath by using read and write count.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: But what does the Entropycoder actually do?

File: litejpeg/core/rle/entropycoder.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            self.sync += [
+             If(get_data[11-i] == 0,
+                size.eq(11-i))
+            ]
+
+        # Connecting the source.data with the output.
+        self.comb += source.data.eq(size)
+
+
+class Entropycoder(PipelinedActor, Module):
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: `EntropyCoder`

File: litejpeg/core/rle/entropycoder.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    """
+    This module will connect the Entropycoder datapath with the input
+    and output either from other modules or from the Test Benches.
+    The input is been taken from the sink and source and is been transferred to
+    the Entropycoder datapath by using read and write count.
+    """
+    def __init__(self):
+
+        # Connecting the module to the input and the output.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: You indenting here is a bit funny...

File: litejpeg/core/rle/entropycoder.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        the next state will be either read or write.
+        Will clear the value of ``write_count`` to be 0.
+        """
+        self.submodules.write_fsm = write_fsm = FSM(reset_state="GET_RESET")
+        write_fsm.act("GET_RESET",
+                      write_clear.eq(1),
+                      If(write_sel != read_sel,
+                         NextState("WRITE_INPUT")))
+
+        """
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py:159]: Why 63?

File: litejpeg/core/rle/rlecore.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+        The input matrix contains two components,
+        the first value of the matrix is called as the DC component and
+        the rest of the values within the matrix are the AC component.
+
+        The DC part is been encoded by subtracting the DC cofficient to that of
+        the DC cofficient of the previous value of the DC matrix.
+
+        The AC cofficients are been encoded by calculating the number of zeros
+        before the non-zero AC cofficient called as the RUnlength.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: Why do we care about the DC/AC coefficients? What are they used for?

File: litejpeg/core/rle/rlecore.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+            self.source.data.eq(self.source_inter.data)
+        ]
+
+
+class Runlength(PipelinedActor, Module):
+    """
+    This module will connect the Rle core datapath with the input
+    and output either from other modules or from the Test Benches.
+    The input is been taken from the sink and source and is been
+    transferred to the RLE core datapath by using read and write count.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: But what is it actually doing / calculating / etc?

File: litejpeg/core/rle/rlecore.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+            self.source_inter.data[16].eq(self.dovalid)
+
+        ]
+
+        self.sync += [
+            self.source.data.eq(self.source_inter.data)
+        ]
+
+
+class Runlength(PipelinedActor, Module):
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: `RunLength`

File: test/common.py [hunk 11]
```diff
@@ -359,4 +360,104 @@ def setdata(self,data):
...
\ No newline at end of file
+
+class RLE:
+    """
+    These class is been created in order to store the value or the matrix which
+    are been used to test the RLE module.
+    The matrix are from github repositories for the purpose of input for
+    testing.
+    """
+    def __init__(self):
```
Comment (mithro) [inline test/common.py]: Where do these values come from? What are they for?

Why is `red_pixels_1`, `self.red_pixels_1` while everything else is not assigned to self?

File: test/common.py [hunk 12]
```diff
@@ -359,4 +360,104 @@ def setdata(self,data):
...
+                          1,  0,  0,  0,  1,  32,   4,  2
+                        ]
+        # Get the output from the reference module.
+        self.output_red_pixels_1 = rle_code(self.red_pixels_1);
+
+    def setdata(self,data):
+        self.data = data
+        for i in range(64):
+            temp=self.data[i]
+            Amplitude = temp%4096
```
Comment (mithro) [inline test/common.py]: `amplitude`

Comment (mithro) [inline test/common.py]: These numbers seem important. Why are you doing a `%4096` and a `%16` below.

File: test/new_rlemain_tb.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
+# This is the module for testing the RLEmain.
+
+# !/usr/bin/env python3
```
Comment (mithro) [inline test/new_rlemain_tb.py]: `# !/usr/bin/env python3` must be the first line in the file.

File: litejpeg/core/rle/entropycoder.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+    """
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the Entropycoder to calculate the bits for the
+    Amplitude to store.
+
+    Parameters:
+    -----------
+    sink : Get the input from the other modules.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: Why does it want the input from the other modules? 

I should be able to understand what the other module this is being connected and why is that a separate module.

File: litejpeg/core/rle/entropycoder.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    source : Give the output to the other modules.
+    input_data : modulus of input of the matrix.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: What matrix? I assume the matrix is coming from somewhere else?

File: litejpeg/core/rle/entropycoder.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+@CEInserter()
+class EntropyDatapath(Module):
+
+    """
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the Entropycoder to calculate the bits for the
+    Amplitude to store.
+
+    Parameters:
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: These are actually `Attributes` not `Parameters`.

Parameters is pretty much a synonym for "argument". For example `def foo(a, b)` has 2 parameters `a` & `b`.

File: litejpeg/core/rle/entropycoder.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+datapath_latency = 3
+
+
+@CEInserter()
+class EntropyDatapath(Module):
+
+    """
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the Entropycoder to calculate the bits for the
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: Entroycoder case is wrong.

File: litejpeg/core/rle/entropycoder.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        of the clock cycle till 63 since we are getting a matrix of 64 blocks
+        hence at write_count equal to 63 it generate a signal saying this to be
+        the last block of the matrix and write the data into the memory
+        as per the data from the ``sink.data`` and when the value reaches 63
+        the state again changes to that of the GET_RESET state.
+        """
+        write_fsm.act("WRITE_INPUT",
+                      sink.ready.eq(1),
+                      If(sink.valid,
+                         If(write_count == 63,
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: `If(write_count == (BLOCK_COUNT-1),` and then define BLOCK_COUNT to be 64 (and use it elsewhere).

File: litejpeg/core/rle/entropycoder.py [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+
+        """
+        WRITE_INPUT State
+
+        Will increament the value of the write_count at every positive edge
+        of the clock cycle till 63 since we are getting a matrix of 64 blocks
+        hence at write_count equal to 63 it generate a signal saying this to be
+        the last block of the matrix and write the data into the memory
+        as per the data from the ``sink.data`` and when the value reaches 63
+        the state again changes to that of the GET_RESET state.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: By changing the code below as I describe, it means that most of this comment can go away. The issue was "why is this counting randomly to 63" -- if it is counting to BLOCK_COUNT it comes really clear what is going on.

File: litejpeg/core/rle/entropycoder.py [hunk 20]
```diff
@@ -0,0 +1,220 @@
+"""
+EntropyCoder Module:
+----------------
+This module is been made in order to calculate the number of bits to
+store the Amplitude.
+
+Parameters:
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: Modules don't have parameters?

File: litejpeg/core/rle/entropycoder.py [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+               ).Elif(write_inc,
+                      write_count.eq(write_count + 1))
+
+        # To combine the datapath into the module
+        self.comb += [
+            self.datapath.sink.data.eq(sink.data + (-2*sink.data[11]*sink.data))
+        ]
+
+        """
+        GET_RESET.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: `GET_RESET` doesn't feel like the right name here but I can't think of a better name... Maybe something like `INIT` or `ON_RESET` or something?

File: litejpeg/core/rle/rlecore.py [hunk 22]
```diff
@@ -0,0 +1,307 @@
+"""
+RLE Core Module:
+----------------
+This module is the core module for the Rle for calculating
+the Amplitude and the Zero_count of the input data.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: I'm still not quite sure what the "Amplitude" is in this context.

File: litejpeg/core/rle/rlecore.py [hunk 23]
```diff
@@ -0,0 +1,307 @@
+"""
+RLE Core Module:
+----------------
+This module is the core module for the Rle for calculating
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: Nit: Use case consistently `RLE` not `Rle`

File: litejpeg/core/rle/rlecore.py (hunk 22 above)
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: `Zero_count` should be *either* `zero_count` or `ZeroCount`

File: litejpeg/core/rle/rlecore.py [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+
+        The AC cofficients are been encoded by calculating the number of zeros
+        before the non-zero AC cofficient called as the Runlength.
+
+        Their is a need of seperate DC/AC cofficients because as the first element
+        of the matrix is the DC cofficient so we already know that the
+        runlength value for DC = 0
+        Therefore while doing huffman encoding we do not want to waste our
+        memory saying the runlength to be zero, that why we take it separate
+        from the rest of the AC cofficients.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: This is a good explanation! It gives me the clear reason why the AC/DC coefficients take different data paths.

File: litejpeg/core/rle/rlecore.py [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+# This is used so that to provide an additional clock attached to the
+# pipeline Actor in the main module.
+@CEInserter()
+
+
+class RLEDatapath(Module):
+
+    def __init__(self):
+
+        """
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: I think this should probably be under the `class RLEDatapath(Module):` ?

File: litejpeg/core/rle/rlecore.py [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+        # Check wheather to start write or not.
+        write_sel = Signal()
+        # To swap the write select.
+        write_swap = Signal()
+        # Check wheather to start read or not.
+        read_sel = Signal(reset=1)
+        # To swap the read_sel.
+        read_swap = Signal()
+
+        # To swap the read and write select whenever required.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: Why would we want to do that?

File: litejpeg/core/rle/rlemain.py [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+RLE main:
+---------
+This module is responsible for dividing the input within the RLEcore and
+the Entrophycoder module. The results are than combined and given
+to the output.
+
+Parameters:
+-----------
+sink : 12 bits
+       Recieves an input of 12 bits either from test bench or from
```
Comment (mithro) [inline litejpeg/core/rle/rlemain.py]: What to the 12 bits of data represent?

File: litejpeg/core/rle/rlemain.py [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+       other modules.
+source : 21 bits
```
Comment (mithro) [inline litejpeg/core/rle/rlemain.py]: What do the 21 bits represent?

File: litejpeg/core/rle/rlemain.py [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+from litejpeg.core.rle.entropycoder import EntropyCoder
+from litejpeg.core.rle.rlecore import RunLength
+
+
+# To keep the output in sync with the input.
+# Addition of a delay of 3 clock cycles.
+datapath_latency = 3
+
+
+class RLEmain(PipelinedActor, Module):
```
Comment (mithro) [inline litejpeg/core/rle/rlemain.py]: `RLEmain` should be `RLEMain` right?

File: test/common.py [hunk 30]
```diff
@@ -359,4 +360,115 @@ def setdata(self,data):
...
         print(self.data)
 
-            
\ No newline at end of file
+
+class RLE:
+    """
+    These class is been created in order to store the value or the matrix which
+    are been used to test the RLE module.
+    The matrix are from github repositories for the purpose of input for
```
Comment (mithro) [inline test/common.py]: Which github repositories? There are plenty out there... How do I check that you haven't made a mistake in the matrixes?

File: test/common.py [hunk 31]
```diff
@@ -359,4 +360,115 @@ def setdata(self,data):
...
+        self.output_red_pixels_1 = rle_code(self.red_pixels_1);
+
+    def setdata(self,data):
+        self.data = data
+        for i in range(64):
+            temp=self.data[i]
+            # Since the data we get contains amplitude, runlength as a single
+            # number therefore in order to extracts the two we take the modulus
+            # by 4096 to get last 12 bits and than shift and than again extracts
+            # the next 4 representing runlength.
```
Comment (mithro) [inline test/common.py]: This doesn't explain why we want to modulus by 4096? Is this the size of the data? Is it got something to do with runlength?

**Review by mithro — COMMENTED** [review]
Another round of comments. You are getting closer, I think we can probably get this merged today!

File: litejpeg/core/rle/entropycoder.py [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the EntropyCoder to calculate the bits for the
+    Amplitude to store.
+
+    Attributes:
+    -----------
+    sink : Get the input from the Entrophycoder.
+    source : Give the output to the Entrophycoder.
+    input_data : for temporary storing the value of input, used
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: input_data / get_data / size are all not attributes. Attributes are things stored on `self`

File: litejpeg/core/rle/entropycoder.py [hunk 33]
```diff
@@ -0,0 +1,223 @@
...
+datapath_latency = 3
+
+
+@CEInserter()
+class EntropyDatapath(Module):
+    """
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the EntropyCoder to calculate the bits for the
+    Amplitude to store.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: Why is amplitude capitalized here?

A better description would be something like `Calculate the bit value for the amplitude using EntropyCoder`.

File: litejpeg/core/rle/entropycoder.py [hunk 34]
```diff
@@ -0,0 +1,223 @@
...
+
+    Attributes:
+    -----------
+    sink : Get the input from the Entrophycoder.
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: Two things here;
 * Capitalization  should always match. If the class is called `EntrophyCoder` you should always refer to it as `EntrophyCoder` never `entrophyCoder` or `Entrophycoder`.
 * "Get the input from the Entrophycoder." is not informative. It should be something like "Accepts data which is encoded in huffman entropy foo bad, such as the output from the EntrophyCoder class." -- The difference is that it tells me *what* the data is.

File: litejpeg/core/rle/entropycoder.py [hunk 35]
```diff
@@ -0,0 +1,223 @@
...
+        # For calculating the size.
+
+        # Storing the size in the input_data.
+        self.sync += input_data.eq(sink.data)
+        # Keep on interating over the input_data and see after shifting
+        # how many bits the input_data value becomes zeros and that is the
+        # number of bits required to store the input_data.
+        for i in range(12):
+            self.sync += get_data[i].eq(input_data >> i)
+        for i in range(12):
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]:
```python
for i in range(11, -1, -1):
```

File: litejpeg/core/rle/entropycoder.py [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+    """
+    This module will connect the Entropycoder datapath with the input
+    and output either from other modules or from the Test Benches.
+    The input is been taken from the sink and source and is been transferred to
+    the Entropycoder datapath by using read and write count.
+    The entrophycoder will extract out the number of bits required to store
+    the amplitude.
+
+    Attributes :
+    ------------
```
Comment (mithro) [inline litejpeg/core/rle/entropycoder.py]: The indenting of the below things in a bit weird...

File: litejpeg/core/rle/rlecore.py [hunk 37]
```diff
@@ -0,0 +1,316 @@
...
+class RLEDatapath(Module):
+    """
+    RLEDatapath Module:
+    -------------------
+    This module is the datapath set for the steps required for
+    the RLE module to take place.
+
+    The input matrix contains two components,
+    the first value of the matrix is called as the DC component and
+    the rest of the values within the matrix are the AC component.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: This is much better description of the input. It explains what the parts are.
```
    The input matrix contains two components,
     * the first value of the matrix is the DC component, and
     * the rest of the values within the matrix are the AC component.
```

File: litejpeg/core/rle/rlecore.py [hunk 38]
```diff
@@ -0,0 +1,316 @@
...
+    before the non-zero AC cofficient called as the Runlength.
+
+    Their is a need of seperate DC/AC cofficients because as the first element
+    of the matrix is the DC cofficient so we already know that the
+    runlength value for DC = 0
+    Therefore while doing huffman encoding we do not want to waste our
+    memory saying the runlength to be zero, that why we take it separate
+    from the rest of the AC cofficients.
+
+    Attributes:
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py:54]: Attributes are only things which are `self.xxx`

File: litejpeg/core/rle/rlecore.py [hunk 39]
```diff
@@ -0,0 +1,316 @@
...
+
+        # Adding PipelineActor to provide additional clock for the module.
+        PipelinedActor.__init__(self, datapath_latency)
+        self.latency = datapath_latency
+
+        # Connecting RLE submodule.
+        self.submodules.datapath = RLEDatapath()
+        self.comb += self.datapath.ce.eq(self.pipe_ce)
+
+        # Intialising the variables.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py]: You don't need to say "initialising the variables" it is clear from the code.

File: litejpeg/core/rle/rlecore.py [hunk 40]
```diff
@@ -0,0 +1,316 @@
...
+    """
+    def __init__(self):
+
+        # Connecting the module to the input and the output.
+        self.sink = sink = stream.Endpoint(
+                               EndpointDescription(block_layout(12)))
+        self.source = source = stream.Endpoint(
+                                   EndpointDescription(block_layout(17)))
+
+        # Adding PipelineActor to provide additional clock for the module.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py:180]: Why do you need an additional clock?

File: litejpeg/core/rle/rlemain.py [hunk 41]
```diff
@@ -0,0 +1,80 @@
...
+
+class RLEMain(PipelinedActor, Module):
+    """
+    RLEMain :
+    ---------
+    This is responsible for combining the outputs of both the EntropyCoder and
+    RLE core as the output will contain the amplitude in the last 12 bits along
+    with the number of bits to store the amplitude as the next 4 bits and
+    finally the RunLength for the next 4 bits.
+    All these information is been synchronized once by the RLEmain and than
```
Comment (mithro) [inline litejpeg/core/rle/rlemain.py]: "and then"

File: test/common.py [hunk 42]
```diff
@@ -359,4 +360,118 @@ def setdata(self,data):
                 self.data[i] = -1*temp;
         print(self.data)
 
-            
\ No newline at end of file
+
+class RLE:
+    """
+    These class is been created in order to store the value or the matrix which
+    are been used to test the RLE module.
```
Comment (mithro) [inline test/common.py]:
```
This class stores the value of the matrixes used to test the RLE module.

These matrixes were taken from [`rle_test_inputs.py` in cfelton's test_jpeg code](https://github.com/cfelton/test_jpeg/blob/master/test/rle_test_inputs.py).

The matrix is an example of what the quantization module might produce.
```

File: test/common.py [hunk 43]
```diff
@@ -359,4 +360,118 @@ def setdata(self,data):
...
+                          1,  0,  0,  0,  0,   0,   0,  1,
+                          1,  0,  0,  0,  1,  32,   4,  2
+                        ]
+        # Get the output from the reference module.
+        self.output_red_pixels_1 = rle_code(self.red_pixels_1);
+
+    def setdata(self,data):
+        self.data = data
+        for i in range(64):
+            temp=self.data[i]
```
Comment (mithro) [inline test/common.py:444]: The data we get contains amplitude and run length as a single number. In order to extract the values we use mod 4096 to get the last 12bits and then shift to extract the next 4 representing the run length.

File: test/new_rlecore_tb.py [hunk 44]
```diff
@@ -0,0 +1,75 @@
...
+from common import *
+
+# Testbanch for the RLEcore module.
+
+"""
+Under this module a matrix containing 64 blocks is been sent as an
+input to the RLE core and the output is been printed and compared
+with the one from the reference modules.
+In this way the result from the RLEcore is been verified.
+"""
```
Comment (mithro) [inline test/new_rlecore_tb.py]: This module takes a matrix containing 64 blocks of XXXX and verifies the RLECore produces the same output as the reference data.

**Review by mithro — COMMENTED** [review]
Almost there for merging this pull request!

File: litejpeg/core/rle/entropycoder.py (hunk 35 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: I don't not able to understand this one.

Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: @mithro

File: test/new_rlecore_tb.py (hunk 44 above)
Comment (ishan98) [reply test/new_rlecore_tb.py]: Done.

File: test/common.py (hunk 43 above)
Comment (ishan98) [reply test/common.py:444]: DONE

File: test/common.py (hunk 42 above)
Comment (ishan98) [reply test/common.py]: DONE

File: litejpeg/core/rle/rlemain.py (hunk 41 above)
Comment (ishan98) [reply litejpeg/core/rle/rlemain.py]: DONE

File: litejpeg/core/rle/rlecore.py (hunk 39 above)
Comment (ishan98) [reply litejpeg/core/rle/rlecore.py]: DONE

File: litejpeg/core/rle/rlecore.py (hunk 40 above)
Comment (ishan98) [reply litejpeg/core/rle/rlecore.py:180]: DONE

File: litejpeg/core/rle/rlecore.py (hunk 38 above)
Comment (ishan98) [reply litejpeg/core/rle/rlecore.py:54]: DONE

File: litejpeg/core/rle/rlecore.py (hunk 37 above)
Comment (ishan98) [reply litejpeg/core/rle/rlecore.py]: DONE

File: litejpeg/core/rle/entropycoder.py (hunk 36 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: DONE

File: litejpeg/core/rle/entropycoder.py (hunk 32 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: DONE

File: litejpeg/core/rle/entropycoder.py (hunk 34 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: DONE

File: litejpeg/core/rle/entropycoder.py (hunk 33 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: DONE

File: litejpeg/core/rle/entropycoder.py (hunk 35 above)
Comment (mithro) [reply litejpeg/core/rle/entropycoder.py]: http://pythoncentral.io/pythons-range-function-explained/

Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: DONE

File: litejpeg/core/rle/entropycoder.py (hunk 19 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py]: DONE

File: litejpeg/core/rle/rlecore.py [hunk 45]
```diff
@@ -0,0 +1,303 @@
...
+        # write path
+
+        # To start the write_count back to 0.
+        write_clear = Signal()
+        # To increment the write_count.
+        write_inc = Signal()
+        # To keep track over which value of the matrix is under process.
+        write_count = Signal(6)
+
+        # For tracking the data adress.
```
Comment (mithro) [inline litejpeg/core/rle/rlecore.py:219]: BTW For migen we normally wrap the code like this;

```python
       self.sync += \
            If(write_clear,
                write_count.eq(0)
            ).Elif(write_inc,
                write_count.eq(write_count + 1)
            )
```
IE
```python
  If(<condition>,
  <4 spaces>statement
  ).Elif(<condition>,
  <4 spaces>statement1,
  <4 spaces>statement2,
  ).Else(
  <4 spaces>statement1,
  <4 spaces>statement2,
  )
```

Comment (ishan98) [reply litejpeg/core/rle/rlecore.py:219]: DONE to most of the places.

File: litejpeg/core/rle/entropycoder.py (hunk 7 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py:159]: DONE replaced with BLOCK_COUNT

File: litejpeg/core/rle/entropycoder.py (hunk 3 above)
Comment (ishan98) [reply litejpeg/core/rle/entropycoder.py:54]: DONE

File: litejpeg/core/rle/rlecore.py (hunk 26 above)
Comment (ishan98) [reply litejpeg/core/rle/rlecore.py]: DONE

**Review by mithro — APPROVED** [review]
LGTM!

Merging.

**Pull request merged**